### PR TITLE
Update config.json schema

### PR DESCRIPTION
**Pull request by N2D4**
----

> [!IMPORTANT]
> Update configuration schema and logic, introducing new RBAC and API Keys schemas, and enhancing validation and normalization processes.
> 
>   - **Schema Updates**:
>     - Introduced new RBAC and API Keys schemas in `schema.ts`.
>     - Updated `branchConfigSchema`, `environmentConfigSchema`, and `organizationConfigSchema` to include new fields.
>     - Added `applyDefaults` function to merge defaults with configs.
>   - **Validation and Normalization**:
>     - Updated `schematicallyValidateAndReturn` in `config.tsx` to handle new schemas.
>     - Modified `normalize` and `override` functions in `format.ts` to support new config structures.
>   - **OAuth and Email Configurations**:
>     - Updated handling of OAuth providers and email configurations in `config.tsx` and `projects.tsx`.
>     - Adjusted `renderedOrganizationConfigToProjectCrud` to reflect new config structures.
>   - **Miscellaneous**:
>     - Updated `README.md` to reflect new config terminology and structure.
>     - Added tests for new and updated functions in `schema-fields.ts` and `objects.tsx`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=stack-auth%2Fstack-auth&utm_source=github&utm_medium=referral)<sup> for 84cd6ca7eb03bf4b098b8b32830bbbe0e7fbe64b. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

**Review by fomalhautb — COMMENTED** [review]
Can you maybe add some more test cases? for example, testing onDotIntoNull=empty. Also I don't see the tests for the final default config overwrite, maybe I am missing it. Otherwise looks good.

File: packages/stack-shared/src/config/schema.ts [hunk 1]
```diff
@@ -1,224 +1,386 @@
...
-    }),
-  ).defined(),
-}).defined();
 
-const branchDomain = yupObject({
-  allowLocalhost: yupBoolean().defined(),
-}).defined();
+const branchAuthSchema = yupObject({
+  allowSignUp: yupBoolean().optional(),
+  allowPasswordSignIn: yupBoolean().optional(),
```
Comment (fomalhautb) [inline packages/stack-shared/src/config/schema.ts]: do we want to nest things like

```tsx
password: {
     isAllowed: boolean
}
```

so for example in the future we might have config like `minPasswordLength`

**Review by fomalhautb — APPROVED** [review]
